### PR TITLE
Doc: Immutable Data FAQ

### PR DESCRIPTION
**Pull request by bundance**
Adds a section of questions on using Immutable Data and Immutable.JS to the FAQ in response to the request in #1785 . Note that although I tried to make this as neutral as possible, this is inevitably an opinionated set of questions and answers (particularly the Best Practices section), so please point out any opinions that are too strong or just plain wrong!

Feedback welcomed.

### CODE REVIEW

**Comment by rmakarov94** [comment]
@bundance Hello, Can you explain me why using your `toJS` hoc give better perf then using `Immutable.toJS()` in `mapDispatchToProps`.
As I understand both of the times Immutable.toJS()  produce new objects and dumb component will rerender.
Thank you

**Comment by markerikson** [comment]
Awright, sitting down to take a look at this.

File: docs/faq/ImmutableData.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+#### Accidental Object Mutation
+With ES6, you can accidentally mutate an object (such as the Redux state tree) quite easily without realising it. For example, updating deeply nested properties, creating a new *reference* to an object instead of a new object, or performing a shallow copy rather than a deep copy, can all lead to inadvertent object mutations, and can trip up even the most experienced JavaScript coder. 
+
+To avoid these issues, ensure you follow the recommended [Immutable Update Patterns for ES6](http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html).
+
+#### Verbose Code
+Updating complex nested state trees can lead to verbose code that is tedious to write and difficult to debug.
+
+#### Poor Performance
+Operating on ES6 objects and arrays in an immutable way can be slow, particularly as your state tree grows larger.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Might want to clarify _why_ this is slow - I believe it has to do with copying properties from object to object, although it would be good to point to some benchmarks on that.

**Comment by bundance** [comment]
Thanks for the review, @markerikson. I'll get to work on the changes you've requested.

File: docs/faq/ImmutableData.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+Immutability can bring increased performance to your app, and leads to simpler programming and debugging, as data that never changes is easier to reason about than data that is free to be changed arbitrarily throughout your app.
+
+In particular, immutability in the context of a Web app enables sophisticated change detection techniques to be implemented simply and cheaply, ensuring the computationally expensive process of updating the DOM occurs only when it absolutely has to (a cornerstone of React’s performance improvements over other libraries).
+<a id="why-is-immutability-required"></a>
+## Why is immutability required in Redux?
+
+In the context of Redux, if your reducers mutate their arguments and directly modify the state tree, you will cause React Redux’s `connect` function to break, as it will be unable to detect changes made to objects and arrays. You MUST therefore use immutable operations in your reducers when using Redux.
+<a id="do-i-have-to-use-immutable-js"></a>
+## Do I have to use Immutable.JS?
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: > First, I think I'd like to see the question rephrased/expanded as "What approaches are there for handling data immutably? Do I have to use Immutable.js?".

I've rephrased the question, but not answered the expanded part (i.e. 'What approaches are there for handling data immutably?'). I'll do that in another PR.

> Second, I would like to see a mention of three possible approaches: "plain JS", immutable update libraries, and Immutable.js, with some relatively short lists of tradeoffs (both pro and con) for each.

Again, I'll do that in a separate PR.

> Third, I'd prefer to refer to "plain" Javascript rather than ES6 specifically, since the object spread operator isn't even final yet (so ES2017 at earliest), and while Object.assign is ES6, the same capability exists with _.extend and its equivalents.

Done.

File: docs/faq/ImmutableData.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
+# Redux FAQ: Immutable Data
+
+## Table of Contents
+- [What are the benefits of Immutability?](#benefits-of-immutability)
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done

File: docs/faq/ImmutableData.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+- [Is Immutable.JS worth the effort?](#is-immutable-js-worth-effort)
+- [What are the Recommended Best Practices for using Immutable.JS with Redux?](#immutable-js-best-practices)
+
+<a id="benefits-of-immutability"></a>
+## What are the benefits of immutability?
+
+Immutability can bring increased performance to your app, and leads to simpler programming and debugging, as data that never changes is easier to reason about than data that is free to be changed arbitrarily throughout your app.
+
+In particular, immutability in the context of a Web app enables sophisticated change detection techniques to be implemented simply and cheaply, ensuring the computationally expensive process of updating the DOM occurs only when it absolutely has to (a cornerstone of React’s performance improvements over other libraries).
+<a id="why-is-immutability-required"></a>
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done

File: docs/faq/ImmutableData.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+## Why is immutability required in Redux?
+
+In the context of Redux, if your reducers mutate their arguments and directly modify the state tree, you will cause React Redux’s `connect` function to break, as it will be unable to detect changes made to objects and arrays. You MUST therefore use immutable operations in your reducers when using Redux.
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: I've expanded this hugely. I must have rewrote it seven times! It's one of those topics that's easy enough to understand, but difficult to explain with writing alone. I think this is why we invented programming languages, as it's definitely easier to understand with code :)

I've got links, but I'll add them to all the documentation I've written in another PR.

File: docs/faq/ImmutableData.md (hunk 1 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: I'll do this in another PR (don't worry, I'm keeping track of the other PRs I'm promising!)

File: docs/faq/ImmutableData.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+#### Verbose Code
+Updating complex nested state trees can lead to verbose code that is tedious to write and difficult to debug.
+
+#### Poor Performance
+Operating on ES6 objects and arrays in an immutable way can be slow, particularly as your state tree grows larger.
+
+**Documentation**
+- [Immutable Update Patterns for ES6](http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html)
+
+<a id="why-use-immutable-library"></a>
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done.

File: docs/faq/ImmutableData.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+## Why should I use an immutable-focused library such as Immutable.JS?
+
+Immutable-focused libraries such as Immutable.js have been designed to overcome the issues with immutability inherent within ES6, providing all the benefits of immutability with the performance your app requires. Whether you choose to use such a library, or stick with ES6, depends on how comfortable you are with adding another dependency to your app, or how sure you are that you can avoid the pitfalls inherent within ES6’s approach to immutability.
+
+Whichever option you choose, make sure you’re familiar with the concepts of [immutability, side effects and mutation](http://redux.js.org/docs/recipes/reducers/PrerequisiteConcepts.html#note-on-immutability-side-effects-and-mutation). In particular, ensure you have a deep understanding of what ES6 does when updating and copying values in order to guard against accidental mutations that will degrade you app’s performance, or break it altogether.
+
+**Documentation**
+- [immutability, side effects and mutation](http://redux.js.org/docs/recipes/reducers/PrerequisiteConcepts.html#note-on-immutability-side-effects-and-mutation)
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: Sorry about that. I'll go through the whole PR to ensure the formatting follows the existing convention once you're happy with the changes.

**Comment by markerikson** [comment]
Great, thanks!  Probably won't have time to look at this until next week, but I'll definitely make sure to come back to this.  Appreciate the time and effort you're putting in!

**Comment by markerikson** [comment]
Just did a quick skim of the updated contents.  I _definitely_ like where you're going with this, but I think there's a few bits that need some further clarification and correction.  Let me go through and mark those up.

File: docs/faq/ImmutableData.md [hunk 8]
```diff
@@ -1,245 +1,145 @@
...
+### How does Redux uses shallow equality checking?
+Redux shallowly compares the value returned from a reducer with the `state` parameter passed into it. If both objects are the same (i.e. the `state` parameter passed in references the same object as that returned by the reducer), then Redux assumes that the state has not been changed and so will leave the state tree unchanged.
 
-Fortunately, this can be easily overcome with the use of a browser extension such as the [Immutable.js Object Formatter](https://chrome.google.com/webstore/detail/immutablejs-object-format/hgldghadipiblonfkkicmgcbbijnpeog), which surfaces your data in Chrome Dev Tools, and hides Immutable.JS’s properties when inspecting your data.
+If, however, the returned value differs from the `state` object passed in (i.e. the `state` parameter passed in references a different object from that returned by the reducer), then Redux assumes a change has been made, and so will apply that change to the state tree.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: I'm pretty sure I know why you're saying this, and I think it's a bit of a misunderstanding.  If you look at the guts of `combineReducers`, it definitely does checks to see if the slice reducers returned a new object reference or not.  If at the end of the loop no changes were detected, it will return the original state object reference instead of the "new" state object.  

However, in the case where the slice reducers are doing direct mutation and _not_ returning new objects, having `combineReducers` return the original state object _will_ result in those values being reflected in the state tree after the reducer logic is done, since the contents of the original state object were modified.

This paragraph probably should be reworded a bit, but I don't have an immediate suggestion for what it should say instead.

File: docs/faq/ImmutableData.md [hunk 9]
```diff
@@ -1,245 +1,145 @@
...
 
-### Breaks object references, causing poor performance
+### How does React-Redux uses shallow equality checking?
+React-Redux shallowly compares the root state with the return values from the `mapStatToProps` function that is passed to its `connect` method in order to trigger the re-rendering of any affected components when the state is mutated.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Typo - should be `mapStateToProps`.

Comment (markerikson) [inline docs/faq/ImmutableData.md]: At least as of 4.x, the wrapper components also do equality checks on the root state object itself.  If that is the same object as before, no further comparisons are done, and `mapState` is not executed.

I haven't dug through 5.x's implementation enough to know how it handles things, but I would assume it does a fairly similar check.

File: docs/faq/ImmutableData.md [hunk 10]
```diff
@@ -1,245 +1,145 @@
...
 
-One of the key advantages of immutability is that it enables shallow equality checking, which dramatically improves performance. If two different variables reference the same immutable object, then a simple equality check of the two variables is enough to determine that they are equal, and that the object they both reference is unchanged. The equality check never has to check the values of any of the object’s properties, as it is, of course, immutable.
+If a shallow check of these values is true, React-Redux assumes the state has not changed, and so will not re-render the connected component.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Maybe reword this a bit?  "If a shallow check of these values shows all of them to be the same as before, React-Redux assumes..."

File: docs/faq/ImmutableData.md [hunk 11]
```diff
@@ -1,245 +1,145 @@
...
+### Why will shallow equality checking not work with mutable objects?
+Shallow equality checking cannot be used to detect if a function mutates an object passed into it if that object is mutable.
 
-This, in turn, causes React’s `shouldComponentUpdate` method (and hence Redux’s `connect` function) to return true every time, causing an unchanged component to be re-rendered every time a change elsewhere in the app is detected, and so  severely degrading performance.
+This is self-evident if you think about it: if two variables reference the same object, they will _always_ be equal, regardless of whether the object changes or not, as they're both referencing the same object. Thus, the following will always return true:
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Think I'd like to see "self-evident" changed for another phrase.  Maybe:  "This is because of how objects and references work: if two..."

File: docs/faq/ImmutableData.md [hunk 12]
```diff
@@ -1,245 +1,145 @@
...
 
-#### Never mix plain JavaScript objects with Immutable.JS  
+### How does shallow equality checking with a mutable object cause problems with Redux?
+Redux holds a reference to the `state` object that’s passed into a reducer function, and compares it with the reference that's returned.
 
-Never let a plain JavaScript object contain Immutable.JS properties. Equally, never let an Immutable.JS object contain a plain JavaScript object.
+Thus, if a reducer mutates any of the properties of the `state` object that's passed in, and then returns a reference to the same _newly mutated_ `state` object, Redux's shallow check of the two references will show them as being equal (although the _values_ within the `state` object have changed, the object reference itself has not - it's still the same `state` object, just with different values.) 
 
-#### Make your entire Redux state tree an Immutable.js object
+_In this case, Redux will not be aware of any changes in your reducer, and the store will not update._
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Per prior comment, I don't think this statement is correct.  Mutations are reflected in the resulting store state, because the old objects were mutated.

File: docs/faq/ImmutableData.md [hunk 13]
```diff
@@ -1,245 +1,145 @@
...
+### How does immutability enable a shallow check to detect object mutations?
+If an object is immutable, any changes that need to be made to it within a function must be made to a _copy_ of the object. 
 
-Something needs to map the Immutable.JS props in your Smart Component to the pure JavaScript props used in your Dumb Component. That something is a Higher Order Component (HOC) that simply takes the Immutable.JS props from your Smart Component, and converts them using `toJS()` to plain JavaScript props, which are then passed to  your Dumb Compnent.
+This mutated copy is a _separate_ object from that passed into the function, and so when it is returned, a shallow check will identify it as being a different object from that passed in, and so return false.
 
-Here is an example of such a HOC:
+### What issues can immutability cause with Redux?
+If you make a copy of the `state` object that's passed into a reducer, and return this copy, then Redux will think that the state has been updated, as you're returning an entirely different object from the `state` object that was passed in.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Maybe clarify this section a bit? I think the key here is that copies should only be made of an object/array _if_ something has actually changed - doing something like `let newState = {...state}` at the start of a reducer could cause excess re-rendering even if you hit the "default" case in a switch statement.

File: docs/faq/ImmutableData.md [hunk 14]
```diff
@@ -1,245 +1,145 @@
...
+### What issues can immutability cause with React-Redux?
+If the properties returned from the `mapStateToProps` function always reference a new object, then the connected component will always be re-rendered, even if the values of that new object have not changed. 
 
-And this is how you would use it in your Smart Component:
+<a id="do-i-have-to-use-immutable-js"></a>
+## What approaches are there for handling data immutably? Do I have to use Immutable.js? 
 
-```
-import { connect } from 'react-redux';
+You do not need to use Immutable.JS with Redux. Plain JavaScript, if written correctly, is perfectly capable of providing immutability without having to use an immutable-focused library. However, guaranteeing immutability with JavaScript is difficult, and it can be easy to mutate an object accidentally, causing bugs in your app that are extremely difficult to locate. For this reason, using a library such as Immutable.JS can significantly improve the reliability of your app, and make your app’s development much easier.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Throw in a mention of "immutable update utility libraries" as an option?

File: docs/faq/ImmutableData.md [hunk 15]
```diff
@@ -1,245 +1,145 @@
...
-By converting Immutable.JS objects to plain JavaScript values within a HOC, we achieve Dumb Component portability, but without the performance hits of using `toJS()` in the Smart Component.
+To avoid these issues, ensure you follow the recommended [immutable update patterns for ES6](http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html).
 
-#### Use the Immutable Object Formatter Chrome Extension to Aid Debugging
+#### Verbose Code
+Updating complex nested state trees can lead to verbose code that is tedious to write and difficult to debug.
 
-Install the [Immutable Object Formatter](https://chrome.google.com/webstore/detail/immutablejs-object-format/hgldghadipiblonfkkicmgcbbijnpeog) , and inspect your Immutable.JS data without seeing the noise of Immutable.JS's own object properties.
+#### Poor Performance
+Operating on JavaScript objects and arrays in an immutable way can be slow, particularly as your state tree grows larger.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Should clarify why/how perf is an issue - as far as I know, it's primarily a problem with copying keys from an old item to a new item.  Great article on the topic at https://medium.com/@dtinth/immutable-js-persistent-data-structures-and-structural-sharing-6d163fbd73d2 that's worth linking here.

File: docs/recipes/UsingImmutableJS.md [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+However, such shallow checking of object equality will not work if your data encapsulated within an Immutable.JS object is itself an object. This is because Immutable.JS’s `toJS()` method, which returns the data contained within an Immutable.js object as a JavaScript value, will create a new object every time it’s called, and so break the reference with the encapsulated data. Accordingly, calling `.toJS()` twice, for example, and assigning the result to two different variables will cause an equality check on those two variables to fail, even though the object values themselves haven’t changed.
+
+This, in turn, causes React’s `shouldComponentUpdate` method (and hence Redux’s `connect` function) to return true every time, causing an unchanged component to be re-rendered every time a change elsewhere in the app is detected, and so  severely degrading performance.
+
+Again, though, this can be prevented quite easily with careful coding, as described in the [Best Practices section](#immutable-js-best-practices) below.
```
Comment (markerikson) [inline docs/recipes/UsingImmutableJS.md]: Repeated usages of the phrases "overcome easily" and "careful coding" are sticking out at me a bit.  Maybe vary the phrasing some?

File: docs/recipes/UsingImmutableJS.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
+# Using Immutable.js with Redux
+## Table of Contents
+- [Why should I use an immutable-focused library such as Immutable.JS?](#why-use-immutable-library)
+- [Why should I choose Immutable.JS as an immutable library?](#why-choose-immutable-js)
+- [What are the issues with using Immutable.JS?](#issues-with-immutable-js)
+- [Is Immutable.JS worth the effort?](#is-immutable-js-worth-effort)
+- [What are the Recommended Best Practices for using Immutable.JS with Redux?](#immutable-js-best-practices)
+
```
Comment (markerikson) [inline docs/recipes/UsingImmutableJS.md]: Can we throw in a caveat paragraph somewhere here towards the front that says something along the lines of: "_If_ you choose to use Immutable.js, here are some (opinionated) best practices that you may find useful"?  I always get nervous about people taking stuff in docs as a "must" or "absolute requirement", and just want to keep it clear that people aren't _required_ to do exactly what this page says.

File: docs/recipes/UsingImmutableJS.md [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+Always.
+
+#### Use Immutable.JS objects in your Smart Components
+
+Smart components that access the store via React Redux’s `connect` function must use the Immutable.JS values returned by your selectors. 
+
+#### Never use `toJS()` in `mapStateToProps`
+
+Converting an Immutable.JS object to a JavaScript object using `toJS()` will return a new object every time. If you use do this in `mapSateToProps`, you will cause the component to believe that the object has changed every time the state tree changes, and so trigger an unnecessary re-render.
```
Comment (markerikson) [inline docs/recipes/UsingImmutableJS.md]: Typo: `mapState`

File: docs/faq/ImmutableData.md [hunk 19]
```diff
@@ -1,245 +1,145 @@
...
+There are four key reasons why Redux requires immutability:
 
-### No ES6 Destructuring or Spread Operators
+- Redux's `combineReducers` utility shallowly checks for reference changes.
+- React-Redux's `connect` method generates components that shallowly check reference changes to the root state with the return values from the `mapStateToProps` function to see if the wrapped components actually need to re-render.
+- Immutable data management ultimately makes data handling safer.
+- Time-travel debugging requires that reducers be pure functions with no side effects, so that you can correctly jump between different states.
 
-Because you have to access your data via Immutable.JS’s own `get()` and `getIn()` methods, you can no longer use ES6’s destructuring and spread operators, making your code more verbose.
+In particular, Redux's use of shallow equality checking requires immutability if the store, and any connected components, are to be updated correctly. To see why, we need to understand the difference between shallow and deep equality checking in JavaScript.
```
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done

File: docs/faq/ImmutableData.md (hunk 8 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: I reworded it...a lot! I closely inspected the Redux code and realised my assumptions of what was going on were wrong. So I've rewritten this bit substantially, which has a knock-on effect of the rest of the discussion on shallow checking and the issues it's caused.

It's added to the depth and length of the FAQ, but I think it's much clearer how and why shallow checking is used in Redux and React-Redux now.

File: docs/faq/ImmutableData.md (hunk 9 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Fixed, with a lot more written about React-Redux and how it uses shallow checking.

File: docs/faq/ImmutableData.md (hunk 10 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done

File: docs/faq/ImmutableData.md (hunk 11 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done.

File: docs/faq/ImmutableData.md (hunk 12 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Fixed (see comment above).

File: docs/faq/ImmutableData.md (hunk 13 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: I've substantially rewritten this bit too, so this should be fixed.

File: docs/faq/ImmutableData.md (hunk 14 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done.

File: docs/faq/ImmutableData.md (hunk 15 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: Done (great article, BTW!)

File: docs/recipes/UsingImmutableJS.md (hunk 17 above)
Comment (bundance) [inline docs/recipes/UsingImmutableJS.md]: Done.

File: docs/recipes/UsingImmutableJS.md (hunk 16 above)
Comment (bundance) [inline docs/recipes/UsingImmutableJS.md]: I've removed them all and been more specific when referring to the best practices.

File: docs/recipes/UsingImmutableJS.md (hunk 18 above)
Comment (bundance) [inline docs/recipes/UsingImmutableJS.md]: Fixed

**Comment by bundance** [comment]
@markerikson All changes now complete. Sorry for taking so long - as you can see, this has been a substantial update, as I examined the Redux and React-Redux code closely and removed some false assumptions I didn't know I had! (Christmas and New Year also didn't help!)

I'm much happier with the end result, though, even if it's led to a deeper and longer FAQ. As usual, please critique and comment as you see fit, as I want this to be a high quality document and am happy to edit/amend as required.

If you're happy with this, I'll add the references and ensure it's formatted correctly, and then hopefully it can be merged in.

**Comment by markerikson** [comment]
No worries.  Sounds like you've spent a lot of time on this, it's not an urgent need with a deadline, and the holidays are definitely a time for doing other stuff anyway :)

I'll take a look at this over the weekend.  Really appreciate the time you've put into this!

**Comment by bundance** [comment]
No problem at all. The second-best way to really learn something is to teach it. The first is to write documentation 😃

File: docs/faq/ImmutableData.md [hunk 20]
```diff
@@ -2,50 +2,154 @@
...
 
-### How does Redux uses shallow equality checking?
-Redux shallowly compares the value returned from a reducer with the `state` parameter passed into it. If both objects are the same (i.e. the `state` parameter passed in references the same object as that returned by the reducer), then Redux assumes that the state has not been changed and so will leave the state tree unchanged.
 
-If, however, the returned value differs from the `state` object passed in (i.e. the `state` parameter passed in references a different object from that returned by the reducer), then Redux assumes a change has been made, and so will apply that change to the state tree.
+<a id="how-redux-uses-shallow-checking"></a>
+### How does Redux use shallow equality checking?
+Redux uses shallow equality checking in its `combineReducers` function, in order to determine whether to return a new mutated root state object, or, if no mutations have been made, the current root state object.  
+
+*If the current root state object is returned by `combineReducers`, the store will not be updated.*
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: I think this line is still an issue.  If no changes were detected, returning the same object either means that nothing really did change, or that mutations occurred.  Either way, the returned object should reflect the current contents of the state.

I think it's probably best to just remove this line entirely.

File: docs/faq/ImmutableData.md [hunk 21]
```diff
@@ -2,50 +2,154 @@
...
+
+Specifically, at each stage of the iteration, `combineReducers` performs a shallow equality check on the current state slice and the state slice returned from the reducer. If the reducer returns a new object, the shallow equality check will fail, and `combineReducers` will set a `hasChanged` flag to true. 
+
+After the iterations have completed, `combineReducers` will check the state of the `hasChanged` flag. If it’s true, the newly-constructed state object will be returned. If it’s false, the _current_ state object is returned.
 
-### How does React-Redux uses shallow equality checking?
-React-Redux shallowly compares the root state with the return values from the `mapStatToProps` function that is passed to its `connect` method in order to trigger the re-rendering of any affected components when the state is mutated.
+This is worth emphasising: *If the reducers all return the same `state` object passed to them, then `combineReducers` will return the current root state object, not the newly updated one, and the store will not be updated.*
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Similarly, the phrase "the store will not be updated" doesn't seem right.

File: docs/faq/ImmutableData.md [hunk 22]
```diff
@@ -61,23 +165,33 @@ param === returnVal;
...
+
+If it does, the shallow equality check that `combineReducers` performs will always pass, as the values of the state slice returned by the reducer may have been mutated, but the object itself has not - it’s still the same object that was passed to the reducer.
 
-Thus, if a reducer mutates any of the properties of the `state` object that's passed in, and then returns a reference to the same _newly mutated_ `state` object, Redux's shallow check of the two references will show them as being equal (although the _values_ within the `state` object have changed, the object reference itself has not - it's still the same `state` object, just with different values.) 
+Accordingly, `combineReducers` will not set its `hasChanged` flag, even though the state has changed. If none of the other reducers return a new, updated state slice, the `hasChanged` flag will remain set to false, causing `combineReducers` to return the _existing_ unmodified state object.
 
-_In this case, Redux will not be aware of any changes in your reducer, and the store will not update._
+*The state, therefore, will not be updated, even though its values should have changed.*
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Still having issues with this phrasing :)

File: docs/faq/ImmutableData.md [hunk 23]
```diff
@@ -291,7 +291,7 @@ Note that, conversely, if the values in your props object refer to mutable objec
 ## What approaches are there for handling data immutably? Do I have to use Immutable.js? 
 You do not need to use Immutable.JS with Redux. Plain JavaScript, if written correctly, is perfectly capable of providing immutability without having to use an immutable-focused library. 
 
-However, guaranteeing immutability with JavaScript is difficult, and it can be easy to mutate an object accidentally, causing bugs in your app that are extremely difficult to locate. For this reason, using a library such as Immutable.JS can significantly improve the reliability of your app, and make your app’s development much easier.
+However, guaranteeing immutability with JavaScript is difficult, and it can be easy to mutate an object accidentally, causing bugs in your app that are extremely difficult to locate. For this reason, using an immutable update utility library such as Immutable.JS can significantly improve the reliability of your app, and make your app’s development much easier.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: I wouldn't call Immutable.js an "immutable update utility library".  It's a full-blown set of specialized data structures.  There's really three categories of JS immutable libs: specialized data structures (Immutable.js, Mori); libs that freeze JS objects (seamless-immutable); libs that abstract out the update process (dot-prop-immutable, etc)

File: docs/faq/ImmutableData.md [hunk 24]
```diff
@@ -309,5 +309,11 @@ Updating complex nested state trees can lead to verbose code that is tedious to
 #### Poor Performance
 Operating on JavaScript objects and arrays in an immutable way can be slow, particularly as your state tree grows larger.
 
+Remember, to change an immutable object, you must mutate a _copy_ of it, and copying large objects can be slow as every property must be copied.
```
Comment (markerikson) [inline docs/faq/ImmutableData.md]: Good info here.

File: docs/recipes/UsingImmutableJS.md [hunk 25]
```diff
@@ -1,4 +1,4 @@
-# Using Immutable.js with Redux
+# # Using Immutable.js with Redux
```
Comment (markerikson) [inline docs/recipes/UsingImmutableJS.md]: Typo: extra '#'

File: docs/recipes/UsingImmutableJS.md [hunk 26]
```diff
@@ -60,49 +60,68 @@ Note that Immutable.JS objects do have a `toJS()` method, which returns the data
...
+When the shallow check fails, React-Redux will cause the component to re-render. Using `toJS()` in `mapStateToProps` in this way, therefore, will always cause the component to re-render, even if the value never changes, impacting heavily on performance.
 
-Again, though, this can be prevented quite easily with careful coding, as described in the [Best Practices section](#immutable-js-best-practices) below.
+This can be prevented by using `toJS()` in a Higher Order Component, as discussed in the [Best Practices section](#immutable-js-best-practices) below.
 
 
 <a id="is-immutable-js-worth-effort"></a>
 ## Is Using Immutable.JS worth the effort?
 
 Yes. Do not underestimate the difficulty of trying to track down a property of your state tree that has been inadvertently mutated. Components will both re-render when they shouldn’t, and refuse to render when they should, and tracking down the bug causing the rendering issue is hard, as the component rendering incorrectly is not necessarily the one whose properties are being accidentally mutated.
```
Comment (markerikson) [inline docs/recipes/UsingImmutableJS.md]: I'd like a _slightly_ more nuanced statement here :)  Maybe along the lines of:

"Frequently, yes.  There are various tradeoffs and opinions to consider, but there are many good reasons to use Immutable.js  Do not..."

File: docs/faq/ImmutableData.md (hunk 20 above)
Comment (bundance) [inline docs/faq/ImmutableData.md]: You're right. I've just re-re-read the code, and my confusion came from the `hasChanged` flag on the last line of `combienReducers`:

```
return hasChanged ? nextState : state
```

If `hasChanged` is false, then the original `state` object is returned. But if the reducer has mutated this object, then of course the store will be updated, as the store maintains a reference to it, and the reducer has directly mutated it. I'm slapping my head that I missed this!

I believe I'm right in saying, though, that if a reducer directly mutates the state object, then any wrapped components will not be re-rendered, as the state may have been mutated, but the state object itself has not changed? This behaviour is caused by the [handleSubsequentCalls() function](https://github.com/reactjs/react-redux/blob/master/src/connect/selectorFactory.js#L76) in React-Redux:

```
const stateChanged = !areStatesEqual(nextState, state)
```

I'll rewrite this bit to focus on this issue, and remove all references to the store not being updated.

Comment (markerikson) [inline docs/faq/ImmutableData.md]: Yup, that's it.  `combineReducers` is trying to micro-optimize things a bit, but either way the return value should contain the latest contents of the store.  And since `connect` is checking for that value to have changed, that's one of the couple places where mutation can cause components to not update (the other being comparison of `mapstate` return values).

**Comment by bundance** [comment]
@markerikson, I've made all the changes asked and rewritten the bit on how shallow checking affects Redux and `combineReducers`. If you're happy with these changes, I'll add the references and ensure the formatting matches the rest of the FAQ, and then hopefully it can be merged in.

**Comment by markerikson** [comment]
Cool, thanks.  I'll take a look at this over the weekend (again :)  ).

**Comment by markerikson** [comment]
Looks pretty good at this point.  Go ahead and make whatever last tweaks you want, and then we can get this in.

**Comment by bundance** [comment]
All finished, @markerikson. I've added loads of references throughout, and hopefully ensured the style matches the rest of the documentation. Let me know if I've missed any references you think are particularly salient (or any other problems!).

**Comment by bundance** [comment]
Not sure if you got my last update, @markerikson - I can imagine you get a ton of them! This PR is now ready for review, with all links and references added. No problem if you're too busy to get round to it - just wanted to make sure you hadn't missed my last update.

**Comment by markerikson** [comment]
@bundance : yeah, sorry - last week's been extremely busy, and this week will be as well.  There's a good chance I won't be able to come back to this until the week of Feb 12.  

I definitely appreciate the time and effort you've put into this, and want to get it in, I just need the time to give it the attention it deserves.

**Comment by bundance** [comment]
No problem at all.

**Comment by markerikson** [comment]
Ooooo-kay.  I didn't _completely_ forget about this, but I did _sorta_ forget about this because I had a whole bunch of other stuff going on the last couple weeks.  Lemme look things over.

**Comment by markerikson** [comment]
Skimmed through the latest diffs, and the current contents.  I'm calling this good to go :)  Let's get it in, and if there's any further changes to be made, we can always make those later.

@bundance , kudos for some excellent info, and thanks for working with me to get that fleshed out.

**Comment by markerikson** [comment]
Now, since we don't have any kind of docs auto-republish step set up, time for me to go re-familiarize myself with the necessary steps and see if I can get this pushed up.

**Comment by markerikson** [comment]
Hmph.  I'm having issues dealing with building and publishing the docs from my box.  I'll have to work with @timdorr to sort this out.

I definitely want to get them republished ASAP, and let people know we've added this section.

**Comment by timdorr** [comment]
It's up! http://redux.js.org/docs/faq/ImmutableData.html
